agent/consul/intention_endpoint.go

```diff
@@ -94,73 +239,74 @@ func (s *Intention) prepareApplyCreate(
 		args.Intention.SourceType = structs.IntentionSourceConsul
 	}
 
-	args.Intention.DefaultNamespaces(entMeta)
-
 	if err := s.validateEnterpriseIntention(args.Intention); err != nil {
-		return nil, err
+		return "", nil, err
 	}
 
 	//nolint:staticcheck
 	if err := args.Intention.Validate(); err != nil {
-		return nil, err
+		return "", nil, err
 	}
 
 	_, configEntry, err := s.srv.fsm.State().ConfigEntry(nil, structs.ServiceIntentions, args.Intention.DestinationName, args.Intention.DestinationEnterpriseMeta())
 	if err != nil {
-		return nil, fmt.Errorf("service-intentions config entry lookup failed: %v", err)
-	} else if configEntry == nil {
-		return nil, nil
+		return "", nil, fmt.Errorf("service-intentions config entry lookup failed: %v", err)
+	}
+
+	if configEntry == nil {
+		return structs.ConfigEntryUpsertCAS, args.Intention.ToConfigEntry(true), nil
+	}
+	prevEntry := configEntry.(*structs.ServiceIntentionsConfigEntry)
+
+	if err := checkLegacyIntentionApplyAllowed(prevEntry); err != nil {
+		return "", nil, err
 	}
 
-	return configEntry.(*structs.ServiceIntentionsConfigEntry), nil
+	upsertEntry := prevEntry.Clone()
+	upsertEntry.Sources = append(upsertEntry.Sources, args.Intention.ToSourceIntention(true))
+
+	// NOTE: if the append of this source causes a duplicate source name the
+	// config entry validation will fail so we don't have to check that
+	// explicitly here.
+
+	return structs.ConfigEntryUpsertCAS, upsertEntry, nil
 }
 
-// prepareApplyUpdateLegacy validates that the requester has permissions on both the updated and existing
-// intention as well as generally validating that the request is well-formed
-//
-// Returns an existing service-intentions config entry for this destination if
-// one exists.
-func (s *Intention) prepareApplyUpdateLegacy(
-	ident structs.ACLIdentity,
+func (s *Intention) computeApplyChangesLegacyUpdate(
+	accessorID string,
 	authz acl.Authorizer,
 	entMeta *structs.EnterpriseMeta,
 	args *structs.IntentionRequest,
-) (*structs.ServiceIntentionsConfigEntry, error) {
-	if !args.Intention.CanWrite(authz) {
```

Comment:
Previously this method did 2 acl checks because destination renames were allowed. I removed one of them, and for consistency with delete-by-id I moved it to after the intention lookup.